README.md
html-to-docx
============

[![NPM Version][npm-image]][npm-url]

html-to-docx is a js library for converting HTML documents to DOCX format supported by Microsoft Word 2007+, LibreOffice Writer, Google Docs, WPS Writer etc.

It was inspired by [html-docx-js] project but mitigates the problem of documents generated being non-compatiable with word processors like Google Docs and libreOffice Writer that doesn't support [altchunks] feature.

html-to-docx earlier used to use [libtidy] to clean up the html before parsing, but had to remove it since it was causing so many dependency issues due to node-gyp.

### Disclaimer

Even though there is an instance of html-to-docx running in production, please ensure that it covers all the cases that you might be encountering usually, since this is not a complete solution.

Currently it doesn't work with browser directly, but it was tested against React.

## Installation

Use the npm to install foobar.

```bash
npm install html-to-docx
```

## Usage

```js
await HTMLtoDOCX(htmlString, headerHTMLString, documentOptions, footerHTMLString)
```

full fledged examples can be found under `example/`

### Parameters

- `htmlString` <[String]> clean html string equivalent of document content.
- `headerHTMLString` <[String]> clean html string equivalent of header. Defaults to `<p></p>` if header flag is `true`.
- `documentOptions` <?[Object]>
  - `orientation` <"portrait"|"landscape"> defines the general orientation of the document. Defaults to portrait.
  - `pageSize` <?[Object]> Defaults to U.S. letter portrait orientation.
    - `width` <[Number]> width of the page for all pages in this section in [TWIP]. Defaults to 12240. Maximum 31680. Supports equivalent measurement in [pixel], [cm] or [inch].
    - `height` <[Number]> height of the page for all pages in this section in [TWIP]. Defaults to 15840. Maximum 31680. Supports equivalent measurement in [pixel], [cm] or [inch].
  - `margins` <?[Object]>
    - `top` <[Number]> distance between the top of the text margins for the main document and the top of the page for all pages in this section in [TWIP]. Defaults to 1440. Supports equivalent measurement in [pixel], [cm] or [inch].
    - `right` <[Number]> distance between the right edge of the page and the right edge of the text extents for this document in [TWIP]. Defaults to 1800. Supports equivalent measurement in [pixel], [cm] or [inch].
    - `bottom` <[Number]> distance between the bottom of text margins for the document and the bottom of the page in [TWIP]. Defaults to 1440. Supports equivalent measurement in [pixel], [cm] or [inch].
    - `left` <[Number]> distance between the left edge of the page and the left edge of the text extents for this document in [TWIP]. Defaults to 1800. Supports equivalent measurement in [pixel], [cm] or [inch].
    - `header` <[Number]> distance from the top edge of the page to the top edge of the header in [TWIP]. Defaults to 720. Supports equivalent measurement in [pixel], [cm] or [inch].
    - `footer` <[Number]> distance from the bottom edge of the page to the bottom edge of the footer in [TWIP]. Defaults to 720. Supports equivalent measurement in [pixel], [cm] or [inch].
    - `gutter` <[Number]> amount of extra space added to the specified margin, above any existing margin values. This setting is typically used when a document is being created for binding in [TWIP]. Defaults to 0. Supports equivalent measurement in [pixel], [cm] or [inch].
  - `title` <?[String]> title of the document.
  - `subject` <?[String]> subject of the document.
  - `creator` <?[String]> creator of the document. Defaults to `html-to-docx`
  - `keywords` <?[Array]<[String]>> keywords associated with the document. Defaults to ['html-to-docx'].
  - `description` <?[String]> description of the document.
  - `lastModifiedBy` <?[String]> last modifier of the document. Defaults to `html-to-docx`.
  - `revision` <?[Number]> revision of the document. Defaults to `1`.
  - `createdAt` <?[Date]> time of creation of the document. Defaults to current time.
  - `modifiedAt` <?[Date]> time of last modification of the document. Defaults to current time.
  - `headerType` <"default"|"first"|"even"> type of header. Defaults to `default`.
  - `header` <?[Boolean]> flag to enable header. Defaults to `false`.
  - `footerType` <"default"|"first"|"even"> type of footer. Defaults to `default`.
  - `footer` <?[Boolean]> flag to enable footer. Defaults to `false`.
  - `font` <?[String]> font name to be used. Defaults to `Times New Roman`.
  - `fontSize` <?[Number]> size of font in HIP(Half of point). Defaults to `22`. Supports equivalent measure in [pt].
  - `complexScriptFontSize` <?[Number]> size of complex script font in HIP(Half of point). Defaults to `22`. Supports equivalent measure in [pt].
  - `table` <?[Object]>
    - `row` <?[Object]>
      - `cantSplit` <?[Boolean]> flag to allow table row to split across pages. Defaults to `false`.
  - `pageNumber` <?[Boolean]> flag to enable page number in footer. Defaults to `false`. Page number works only if footer flag is set as `true`.
  - `skipFirstHeaderFooter` <?[Boolean]> flag to skip first page header and footer. Defaults to `false`.
  - `lineNumber` <?[Boolean]> flag to enable line numbering. Defaults to `false`.
  - `lineNumberOptions` <?[Object]>
    - `start` <[Number]> start of the numbering - 1. Defaults to `0`.
    - `countBy` <[Number]> skip numbering in how many lines in between + 1. Defaults to `1`.
    - `restart` <"continuous"|"newPage"|"newSection"> numbering restart strategy. Defaults to `continuous`.
  - `numbering` <?[Object]>
    - `defaultOrderedListStyleType` <?[String]> default ordered list style type. Defaults to `decimal`.
  - `decodeUnicode` <?[Boolean]> flag to enable unicode decoding of header, body and footer. Defaults to `false`.
  - `lang` <?[String]> language localization code for spell checker to work properly. Defaults to `en-US`.
- `footerHTMLString` <[String]> clean html string equivalent of footer. Defaults to `<p></p>` if footer flag is `true`.

### Returns

<[Promise]<[Buffer]|[Blob]>>

## Notes

Currently page break can be implemented by having div with classname "page-break" or style "page-break-after" despite the values of the "page-break-after", and contents inside the div element will be ignored. `<div class="page-break" style="page-break-after: always;"></div>`

<br>

CSS list-style-type for `<ol>` element are now supported. Just do something like this in the HTML:
```
  <ol style="list-style-type:lower-alpha;">
    <li>List item</li>
    ...
  </ol>
```
List of supported list-style-types:
- upper-alpha, will result in `A. List item`
- lower-alpha, will result in `a. List item`
- upper-roman, will result in `I. List item`
- lower-roman, will result in `i. List item`
- lower-alpha-bracket-end, will result in `a) List item`
- decimal-bracket-end, will result in `1) List item`
- decimal-bracket, will result in `(1) List item`
- decimal, **(the default)** will result in `1. List item`

Also you could add attribute `data-start="n"` to start the numbering from the n-th.

`<ol data-start="2">` will start the numbering from ( B. b. II. ii. 2. )

## Contributing

Pull requests are welcome. For major changes, please open an issue first to discuss what you would like to change.

Please make sure to branch new branches off of develop for contribution.

## Support

[!["Buy Me A Coffee"](https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png)](https://www.buymeacoffee.com/privateOmega)

## License

MIT

[npm-image]: https://img.shields.io/npm/v/html-to-docx.svg
[npm-url]: https://npmjs.org/package/html-to-docx
[html-docx-js]: https://github.com/evidenceprime/html-docx-js "html-docx-js"
[altchunks]: https://docs.microsoft.com/en-us/dotnet/api/documentformat.openxml.wordprocessing.altchunk?view=openxml-2.8.1 "altchunks"
[libtidy]: https://github.com/jure/node-libtidy "libtidy"
[String]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#String_type "String"
[Object]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object "Object"
[Number]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Number_type "Number"
[TWIP]: https://en.wikipedia.org/wiki/Twip "TWIP"
[Array]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array "Array"
[Date]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date "Date"
[Boolean]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Data_structures#Boolean_type "Boolean"
[Promise]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise "Promise"
[Buffer]: https://nodejs.org/api/buffer.html#buffer_buffer "Buffer"
[Blob]: https://developer.mozilla.org/en-US/docs/Web/API/Blob "Blob"
[pixel]: https://en.wikipedia.org/wiki/Pixel#:~:text=Pixels%2C%20abbreviated%20as%20%22px%22,what%20screen%20resolution%20views%20it. "pixel"
[cm]: https://en.wikipedia.org/wiki/Centimetre "cm"
[inch]: https://en.wikipedia.org/wiki/Inch "inch"
[pt]: https://en.wikipedia.org/wiki/Point_(typography) "pt"

## Contributors

<a href="https://github.com/privateomega/html-to-docx/graphs/contributors">
  <img src="https://contrib.rocks/image?repo=privateomega/html-to-docx" />
</a>

Made with [contrib.rocks](https://contrib.rocks).

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.9.0 (2023-06-26)


### Features

* abstracted conversion using docxDocument class ([c625a01](https://github.com/privateOmega/html-to-docx/commit/c625a0181a6c328c0319b579fa1173192dff1187))
* add pageSize as an optional additional document option ([ed9aeda](https://github.com/privateOmega/html-to-docx/commit/ed9aedaf1ad462bd0ce5d284fd915d24d280c428))
* add support for code tag ([70a9485](https://github.com/privateOmega/html-to-docx/commit/70a948503d6b8c11b3f46141578047361c0bba87))
* add support for lists within table cells ([b7d5ce7](https://github.com/privateOmega/html-to-docx/commit/b7d5ce7c6381c8e5ea79e537ee78768c9fefdbb5))
* add support for nested images ([882014f](https://github.com/privateOmega/html-to-docx/commit/882014f4017359139dbbfd5f7c8c14cae48f55bf))
* add support for pre tag ([c43ed9e](https://github.com/privateOmega/html-to-docx/commit/c43ed9e3bef91d1c95eddb8779df8e2d2e7f151b))
* add support for urls in image src ([60c7e5b](https://github.com/privateOmega/html-to-docx/commit/60c7e5bcefaaee26068206087be0345d0c79b988))
* add valid url regex and util method ([c5020ce](https://github.com/privateOmega/html-to-docx/commit/c5020ce220e324c7ba0d48976998604162b3a73f))
* added b tag support ([f867abd](https://github.com/privateOmega/html-to-docx/commit/f867abd41c6bc85bbba207a27c58d441f1a2b532))
* added builder methods for images ([9e2720f](https://github.com/privateOmega/html-to-docx/commit/9e2720f261a46701c8a2581aadafa9b60e6cee6b))
* added css border string parser ([15562e8](https://github.com/privateOmega/html-to-docx/commit/15562e817732d1d85b99fd33921694f2e3ad3ad7))
* added css color string ([cb0db2f](https://github.com/privateOmega/html-to-docx/commit/cb0db2ff2d3f2f66df823dbafbc5603030241bc3))
* added document file render helper ([6dd9c3a](https://github.com/privateOmega/html-to-docx/commit/6dd9c3a01f5fceab78404d8ebddb848fb91c933c))
* added eip conversions ([9d6e317](https://github.com/privateOmega/html-to-docx/commit/9d6e3171c0b0f9d71f776762357d4a329778cedb))
* added em tag support ([6a06265](https://github.com/privateOmega/html-to-docx/commit/6a06265f724a611b50144cb988e576bc4e40b4d4))
* added escape-html ([1a231d5](https://github.com/privateOmega/html-to-docx/commit/1a231d5dde3e6f9b5a23f248e19063191c07e54f))
* added font size support ([0f27c60](https://github.com/privateOmega/html-to-docx/commit/0f27c609baa5b9488bc195dff1c060bcc04bbf2d))
* added font support in styles ([18b3281](https://github.com/privateOmega/html-to-docx/commit/18b3281ac3f91e5c1905efa0487354ff78badec2))
* added font table ([0903d6b](https://github.com/privateOmega/html-to-docx/commit/0903d6b98fae6dc378cdeafdadd80a86501c9959))
* added footer support ([24f60df](https://github.com/privateOmega/html-to-docx/commit/24f60df94d755689493b761340d16d090c8b1b16))
* added generic rels xml string ([7d2ea84](https://github.com/privateOmega/html-to-docx/commit/7d2ea8404093fd99ce5e6bccaedaf3603830f574))
* added header generation ([25fb44f](https://github.com/privateOmega/html-to-docx/commit/25fb44f945df3fdc5f37d619b3de3ebe68b84cd6))
* added header relationship support ([cc08355](https://github.com/privateOmega/html-to-docx/commit/cc083553e28a3ed88da3cc27ef0de0cbb26350b3))
* added heading sizes ([bb18e72](https://github.com/privateOmega/html-to-docx/commit/bb18e724c42b0c4581722b2899d5ff808c1495c4))
* added headings support ([fd489ee](https://github.com/privateOmega/html-to-docx/commit/fd489eeebfeedc7d05991f9366aeae2adc49fd6f))
* added highlight support ([6159925](https://github.com/privateOmega/html-to-docx/commit/6159925495b74ab254cd7dc5628526d531595a92))
* added horizontal text alignment ([d29669f](https://github.com/privateOmega/html-to-docx/commit/d29669ffdb0d63b7bdbbe09c6bca990e4c28cfb8))
* added hsl conversion support ([153fa43](https://github.com/privateOmega/html-to-docx/commit/153fa43f84c640085f45823bc2054b24c28023d0))
* added hyperlink styling ([6c3f1bd](https://github.com/privateOmega/html-to-docx/commit/6c3f1bd92f5c7b4ddf74beb1fc3f4e2e6b4762f5))
* added hyperlinks support ([3560ce9](https://github.com/privateOmega/html-to-docx/commit/3560ce9f23fa8f590aa340302bf0059c8dfb6d5f))
* added indentation support ([9c1c3a4](https://github.com/privateOmega/html-to-docx/commit/9c1c3a4d539f4dfeda2ab6484d69236a91ec5ae1))
* added ins tag support ([6d64908](https://github.com/privateOmega/html-to-docx/commit/6d64908858dac290aa34421c236bdaf2d8ef07a7))
* added line height support ([3d0ea2f](https://github.com/privateOmega/html-to-docx/commit/3d0ea2fe56d13893e3c5cd0e4a35e7b26b7c1d0a))
* added line numbers ([f87202c](https://github.com/privateOmega/html-to-docx/commit/f87202c4d4cd534e830d3a59d33b9cc27cf6d654))
* added linebreak support ([57c054c](https://github.com/privateOmega/html-to-docx/commit/57c054cd65f49d7c4244272af0117f2c141a8bc7))
* added method to archive images with other files ([b6da74b](https://github.com/privateOmega/html-to-docx/commit/b6da74be10be03d689ca044f3f95dd724a3a29b6))
* added minimum width support to tables ([b10d820](https://github.com/privateOmega/html-to-docx/commit/b10d820061b10531aa027fde304fbe3ceac849d5))
* added more unit converters ([8f78c52](https://github.com/privateOmega/html-to-docx/commit/8f78c5241cf33d471c8b08e3f941f401d6a50d7b))
* added more xml builder methods ([ffc584b](https://github.com/privateOmega/html-to-docx/commit/ffc584bed7ab434431999517a3308483ba99489a))
* added more xml statment builder methods ([337e530](https://github.com/privateOmega/html-to-docx/commit/337e5305aa8768b6507323bec2279d557a35b67b))
* added other measure units for margins and fonts ([1ae584a](https://github.com/privateOmega/html-to-docx/commit/1ae584a1b0a5350943e10c0d129402b843d7b9a2))
* added page break support ([085ed2a](https://github.com/privateOmega/html-to-docx/commit/085ed2a2cb439ee2a4189b3664deca047926672b))

src/constants.js
import { cloneDeep } from 'lodash';

const applicationName = 'html-to-docx';
const defaultOrientation = 'portrait';
const landscapeWidth = 15840;
const landscapeHeight = 12240;
const landscapeMargins = {
  top: 1800,
  right: 1440,
  bottom: 1800,
  left: 1440,
  header: 720,
  footer: 720,
  gutter: 0,
};
const portraitMargins = {
  top: 1440,
  right: 1800,
  bottom: 1440,
  left: 1800,
  header: 720,
  footer: 720,
  gutter: 0,
};
const defaultFont = 'Times New Roman';
const defaultFontSize = 22;
const defaultLang = 'en-US';
const defaultDocumentOptions = {
  orientation: defaultOrientation,
  margins: cloneDeep(portraitMargins),
  title: '',
  subject: '',
  creator: applicationName,
  keywords: [applicationName],
  description: '',
  lastModifiedBy: applicationName,
  revision: 1,
  createdAt: new Date(),
  modifiedAt: new Date(),
  headerType: 'default',
  header: false,
  footerType: 'default',
  footer: false,
  font: defaultFont,
  fontSize: defaultFontSize,
  complexScriptFontSize: defaultFontSize,
  table: {
    row: {
      cantSplit: false,
    },
  },
  pageSize: {
    width: landscapeHeight,
    height: landscapeWidth,
  },
  pageNumber: false,
  skipFirstHeaderFooter: false,
  lineNumber: false,
  lineNumberOptions: {
    countBy: 1,
    start: 0,
    restart: 'continuous',
  },
  numbering: {
    defaultOrderedListStyleType: 'decimal',
  },
  decodeUnicode: false,
  defaultLang,
};
const defaultHTMLString = '<p></p>';
const relsFolderName = '_rels';
const headerFileName = 'header1';
const footerFileName = 'footer1';
const themeFileName = 'theme1';
const documentFileName = 'document';
const headerType = 'header';
const footerType = 'footer';
const themeType = 'theme';
const hyperlinkType = 'hyperlink';
const imageType = 'image';
const internalRelationship = 'Internal';
const wordFolder = 'word';
const themeFolder = 'theme';
const paragraphBordersObject = {
  top: {
    size: 0,
    spacing: 3,
    color: 'FFFFFF',
  },
  left: {
    size: 0,
    spacing: 3,
    color: 'FFFFFF',
  },
  bottom: {
    size: 0,
    spacing: 3,
    color: 'FFFFFF',
  },
  right: {
    size: 0,
    spacing: 3,
    color: 'FFFFFF',
  },
};
const colorlessColors = ['transparent', 'auto'];
const verticalAlignValues = ['top', 'middle', 'bottom'];

export {
  defaultDocumentOptions,
  defaultHTMLString,
  relsFolderName,
  headerFileName,
  footerFileName,
  themeFileName,
  documentFileName,
  headerType,
  footerType,
  themeType,
  internalRelationship,
  wordFolder,
  themeFolder,
  landscapeMargins,
  portraitMargins,
  defaultOrientation,
  landscapeWidth,
  landscapeHeight,
  applicationName,
  defaultFont,
  defaultFontSize,
  hyperlinkType,
  imageType,
  paragraphBordersObject,
  colorlessColors,
  verticalAlignValues,
  defaultLang,
};
